### ISSUE TITLE
Date transformation during invoice generation may lead to not finding catalog

### ISSUE THREAD

**Issue by sbrossie** [killbill/killbill#760]
Let's assume the following scenario.

1. Create a tenant and add one `SimplePlan` -> will create a catalog for that tenant
2. Create an account
3. Create a subscription (no `TRIAL`).

We expect invoice generation to happen, but it fails with following exception:

```
org.killbill.billing.invoice.api.InvoiceApiException: There is no catalog version that applies for the given date 'Thu Jun 22 19:06:08 GMT 2017'
	at org.killbill.billing.invoice.generator.FixedAndRecurringInvoiceItemGenerator.processRecurringEvent(FixedAndRecurringInvoiceItemGenerator.java:259)
	at org.killbill.billing.invoice.generator.FixedAndRecurringInvoiceItemGenerator.processRecurringBillingEvents(FixedAndRecurringInvoiceItemGenerator.java:148)
	at org.killbill.billing.invoice.generator.FixedAndRecurringInvoiceItemGenerator.generateItems(FixedAndRecurringInvoiceItemGenerator.java:108)
	at org.killbill.billing.invoice.generator.DefaultInvoiceGenerator.generateInvoice(DefaultInvoiceGenerator.java:84)
	at org.killbill.billing.invoice.InvoiceDispatcher.processAccountWithLockAndInputTargetDate(InvoiceDispatcher.java:385)
	at org.killbill.billing.invoice.InvoiceDispatcher.processAccountWithLock(InvoiceDispatcher.java:288)
	at org.killbill.billing.invoice.InvoiceDispatcher.processAccount(InvoiceDispatcher.java:251)
	at org.killbill.billing.invoice.InvoiceDispatcher.processAccountFromNotificationOrBusEvent(InvoiceDispatcher.java:228)
	at org.killbill.billing.invoice.InvoiceDispatcher.processSubscriptionInternal(InvoiceDispatcher.java:210)
	at org.killbill.billing.invoice.InvoiceDispatcher.processSubscriptionForInvoiceGeneration(InvoiceDispatcher.java:185)
	at org.killbill.billing.invoice.InvoiceDispatcher.processSubscriptionForInvoiceGeneration(InvoiceDispatcher.java:181)
	at org.killbill.billing.invoice.InvoiceListener.handleSubscriptionTransition(InvoiceListener.java:80)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.google.common.eventbus.Subscriber.invokeSubscriberMethod(Subscriber.java:91)
	at com.google.common.eventbus.Subscriber$1.run(Subscriber.java:76)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:435)
	at com.google.common.eventbus.Subscriber.dispatchEvent(Subscriber.java:71)
	at com.google.common.eventbus.Dispatcher$ImmediateDispatcher.dispatch(Dispatcher.java:200)
	at com.google.common.eventbus.EventBusThatThrowsException.postWithException(EventBusThatThrowsException.java:65)
	at org.killbill.bus.DefaultPersistentBus.dispatchBusEventWithMetrics(DefaultPersistentBus.java:314)
	at org.killbill.bus.dispatching.BusCallableCallback.dispatch(BusCallableCallback.java:40)
	at org.killbill.bus.dispatching.BusCallableCallback.dispatch(BusCallableCallback.java:29)
	at org.killbill.queue.dispatching.Dispatcher$CallableQueue.call(Dispatcher.java:103)
	at org.killbill.queue.dispatching.Dispatcher$CallableQueue.call(Dispatcher.java:86)
	at org.killbill.commons.concurrent.WrappedCallable.call(WrappedCallable.java:42)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.killbill.commons.concurrent.WrappedRunnable.run(WrappedRunnable.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.killbill.billing.catalog.api.CatalogApiException: There is no catalog version that applies for the given date 'Thu Jun 22 19:06:08 GMT 2017'
	at org.killbill.billing.catalog.VersionedCatalog.indexOfVersionForDate(VersionedCatalog.java:125)
	at org.killbill.billing.catalog.VersionedCatalog.versionsBeforeDate(VersionedCatalog.java:111)
	at org.killbill.billing.catalog.VersionedCatalog.findCatalogPlanEntry(VersionedCatalog.java:158)
	at org.killbill.billing.catalog.VersionedCatalog.findPlan(VersionedCatalog.java:310)
	at org.killbill.billing.catalog.VersionedCatalog.findPhase(VersionedCatalog.java:341)
	at org.killbill.billing.junction.plumbing.billing.DefaultBillingEvent.getRecurringPrice(DefaultBillingEvent.java:208)
	at org.killbill.billing.invoice.generator.FixedAndRecurringInvoiceItemGenerator.processRecurringEvent(FixedAndRecurringInvoiceItemGenerator.java:234)
	... 32 common frames omitted
```

**Comment by sbrossie** [comment]
The issue has to do with the date transformation [here](https://github.com/killbill/killbill/blob/killbill-0.18.9/invoice/src/main/java/org/killbill/billing/invoice/generator/FixedAndRecurringInvoiceItemGenerator.java#L234).

We end up creating an ``effectiveDate` that may be prior to the catalog `effectiveDate` and that breaks;

**Comment by sbrossie** [comment]
Bug was fixed in `0.18.10`, but i think we should re-open to also fix the logic in [VersionedCatalog #findCatalogPlanEntry](https://github.com/killbill/killbill/blob/killbill-0.18.10/catalog/src/main/java/org/killbill/billing/catalog/VersionedCatalog.java#L185):

If the `subscriptionStartDate` is incorrect -- approximated from a `LocalDate` -> `DateTime` @see killbill/killbill-analytics-plugin#70 -- because the api does not allow to retrieve correct date, you may end up with a `subscriptionStartDate` prior the `effectiveDate` of earliest catalog version and throw an exception. In this case i suggest we also default to first version of catalog.

@pierre Any thoughts?

**Comment by sbrossie** [comment]
Also seems like we have a report of the issue from within KB core -- and through a plugin. See comment [here](https://github.com/killbill/killbill/issues/56#issuecomment-320811037), so re-opening the issue.
